### PR TITLE
[runtime] Use MonoError for mono_string_new_size

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
Depends on https://github.com/mono/mono/pull/2472

File: mono/metadata/object.c [hunk 1]
```diff
@@ -5007,9 +5012,11 @@ mono_string_new_utf16 (MonoDomain *domain, const guint16 *text, gint32 len)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
+	MonoError error;
 	MonoString *s;
 	
-	s = mono_string_new_size (domain, len);
+	s = mono_string_new_size_checked (domain, len, &error);
+	mono_error_raise_exception (&error); /* FIXME don't raise here */
 	g_assert (s != NULL);
```
Comment (kumpera) [inline mono/metadata/object.c]: I guess this assert could go

File: mono/metadata/object.c [hunk 2]
```diff
@@ -5029,20 +5036,22 @@ mono_string_new_utf32 (MonoDomain *domain, const mono_unichar4 *text, gint32 len
...
-		g_error_free (error);
+	if (gerror)
+		g_error_free (gerror);
 
 	while (utf16_output [utf16_len]) utf16_len++;
 	
-	s = mono_string_new_size (domain, utf16_len);
+	s = mono_string_new_size_checked (domain, utf16_len, &error);
+	mono_error_raise_exception (&error); /* FIXME don't raise here */
 	g_assert (s != NULL);
```
Comment (kumpera) [inline mono/metadata/object.c]: the != null asserts can go now that we ensure that to happen

File: mono/metadata/object.c [hunk 3]
```diff
@@ -5062,15 +5071,29 @@ mono_string_new_utf32 (MonoDomain *domain, const mono_unichar4 *text, gint32 len
...
 	/* check for overflow */
-	if (len < 0 || len > ((SIZE_MAX - G_STRUCT_OFFSET (MonoString, chars) - 8) / 2))
-		mono_gc_out_of_memory (-1);
+	if (len < 0 || len > ((SIZE_MAX - G_STRUCT_OFFSET (MonoString, chars) - 8) / 2)) {
+		mono_error_set_out_of_memory (error, "Could not allocate %i bytes", -1);
+		return NULL;
+	}
 
 	size = (G_STRUCT_OFFSET (MonoString, chars) + (((size_t)len + 1) * 2));
 	g_assert (size > 0);
```
Comment (kumpera) [inline mono/metadata/object.c]: This is a good time to promote this assert into proper error handling.

Comment (ludovic-henry) [inline mono/metadata/object.c]: We already check that size would not overflow. So if it actually does, we should assert.

**Pull request merged**